File: packages/guards/src/context-poisoning.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ContextPoisoningOptions {
  action: 'block' | 'warn';
}

const POISONING_PATTERNS: RegExp[] = [
  /from\s+now\s+on/i,
  /for\s+all\s+future\s+responses/i,
  /remember\s+this\s+for\s+later/i,
  /always\s+respond\s+with/i,
  /never\s+mention/i,
  /in\s+every\s+response/i,
  /update\s+your\s+instructions/i,
  /add\s+to\s+your\s+rules/i,
  /new\s+permanent\s+rule/i,
  /going\s+forward/i,
  /from\s+this\s+point\s+on/i,
  /\bhenceforth\b/i,
  /embed\s+this\s+in\s+your\s+context/i,
  /store\s+this\s+instruction/i,
  /save\s+this\s+rule/i,
];

const IMPERATIVE_RE = /\b(always|never|must|shall)\b/i;

export function contextPoisoning(options: ContextPoisoningOptions): Guard {
  return {
    name: 'context-poisoning',
    version: '0.1.0',
    description: 'Detects injected instructions that persist across conversation turns',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pat of POISONING_PATTERNS) {
        if (pat.test(text)) {
          matched.push(pat.source);
        }
      }

      const hasImperative = IMPERATIVE_RE.test(text);
      const triggered = matched.length > 0 && hasImperative;

      return {
        guardName: 'context-poisoning',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { matchedPatterns: matched.length, hasImperative }
          : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/context_poisoning.py
"""Detects injected instructions that persist across conversation turns."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

POISONING_PATTERNS = [
    re.compile(r"from\s+now\s+on", re.I),
    re.compile(r"for\s+all\s+future\s+responses", re.I),
    re.compile(r"remember\s+this\s+for\s+later", re.I),
    re.compile(r"always\s+respond\s+with", re.I),
    re.compile(r"never\s+mention", re.I),
    re.compile(r"in\s+every\s+response", re.I),
    re.compile(r"update\s+your\s+instructions", re.I),
    re.compile(r"add\s+to\s+your\s+rules", re.I),
    re.compile(r"new\s+permanent\s+rule", re.I),
    re.compile(r"going\s+forward", re.I),
    re.compile(r"from\s+this\s+point\s+on", re.I),
    re.compile(r"\bhenceforth\b", re.I),
    re.compile(r"embed\s+this\s+in\s+your\s+context", re.I),
    re.compile(r"store\s+this\s+instruction", re.I),
    re.compile(r"save\s+this\s+rule", re.I),
]

IMPERATIVE_RE = re.compile(r"\b(always|never|must|shall)\b", re.I)


class _ContextPoisoning:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "context-poisoning"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched = [p.pattern for p in POISONING_PATTERNS if p.search(text)]
        has_imperative = bool(IMPERATIVE_RE.search(text))
        triggered = len(matched) > 0 and has_imperative
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="context-poisoning",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Context poisoning detected: {len(matched)} persistent-instruction pattern(s) with imperative" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched_patterns": len(matched), "has_imperative": has_imperative} if triggered else None,
        )


def context_poisoning(*, action: str = "block") -> _ContextPoisoning:
    return _ContextPoisoning(action=action)
